File: support/android/apk/servoapp/src/main/java/org/servo/servoshell/HistoryEntry.java
package org.servo.servoshell;

import org.json.JSONException;
import org.json.JSONObject;

public class HistoryEntry {
    private long timestamp;
    private String url;
    private String title;

    public HistoryEntry(long timestamp, String url, String title) {
        this.timestamp = timestamp;
        this.url = url;
        this.title = title;
    }

    public long getTimestamp() {
        return timestamp;
    }

    public String getUrl() {
        return url;
    }

    public String getTitle() {
        return title;
    }

    // Convert to JSON
    public JSONObject toJSON() throws JSONException {
        JSONObject json = new JSONObject();
        json.put("timestamp", timestamp);
        json.put("url", url);
        json.put("title", title != null ? title : "");
        return json;
    }

    // Create from JSON
    public static HistoryEntry fromJSON(JSONObject json) throws JSONException {
        long timestamp = json.getLong("timestamp");
        String url = json.getString("url");
        String title = json.optString("title", "");
        return new HistoryEntry(timestamp, url, title);
    }
}

File: support/android/apk/servoapp/src/main/java/org/servo/servoshell/HistoryEntry.kt
package org.servo.servoshell

import org.json.JSONException
import org.json.JSONObject

class HistoryEntry(val timestamp: Long, val url: String, val title: String?) {
    @Throws(JSONException::class)
    fun toJSON() = JSONObject().apply {
        put("timestamp", timestamp)
        put("url", url)
        put("title", title.orEmpty())
    }

    companion object {
        @JvmStatic
        @Throws(JSONException::class)
        fun fromJSON(json: JSONObject) = HistoryEntry(
            timestamp = json.getLong("timestamp"),
            url = json.getString("url"),
            title = json.optString("title", ""),
        )
    }
}
